python/cudf/cudf/_lib/pylibcudf/interop.pyx

```diff
@@ -1,34 +1,211 @@
+    """
+    name: str = ""
+    children_meta: list[ColumnMetadata] = field(default_factory=list)
+
+
+@singledispatch
+def from_arrow(pyarrow_object, *, DataType data_type=None):
+    """Create a cudf object from a pyarrow object.
+
+    Parameters
+    ----------
+    pyarrow_object : Union[pyarrow.Array, pyarrow.Table, pyarrow.Scalar]
+        The PyArrow object to convert.
+
+    Returns
+    -------
+    Union[Table, Scalar]
+        The converted object of type corresponding to the input type in cudf.
+    """
+    raise TypeError("from_arrow only accepts Table and Scalar objects")
+
+
+@from_arrow.register(pa.Table)
+def _from_arrow_table(pyarrow_object, *, DataType data_type=None):
+    if data_type is not None:
+        raise ValueError("data_type may not be passed for tables")
+    cdef shared_ptr[pa.CTable] arrow_table = pa.pyarrow_unwrap_table(pyarrow_object)
+
+    cdef unique_ptr[table] c_result
+    with nogil:
+        c_result = move(cpp_from_arrow(dereference(arrow_table)))
+
+    return Table.from_libcudf(move(c_result))
+
+
+@from_arrow.register(pa.Scalar)
+def _from_arrow_scalar(pyarrow_object, *, DataType data_type=None):
+    cdef shared_ptr[pa.CScalar] arrow_scalar = pa.pyarrow_unwrap_scalar(pyarrow_object)
+
+    cdef unique_ptr[scalar] c_result
+    with nogil:
+        c_result = move(cpp_from_arrow(dereference(arrow_scalar)))
+
+    cdef Scalar result = Scalar.from_libcudf(move(c_result))
+
+    if result.type().id() != type_id.DECIMAL128:
+        if data_type is not None:
+            raise ValueError(
+                "dtype may not be passed for non-decimal types"
+            )
+        return result
+
+    if data_type is None:
+        raise ValueError(
+            "Decimal scalars must be constructed with a dtype"
+        )
+
+    cdef type_id tid = data_type.id()
+
+    if tid == type_id.DECIMAL32:
+        result.c_obj.reset(
+            new fixed_point_scalar[decimal32](
+                (
+                    <fixed_point_scalar[decimal128]*> result.c_obj.get()
+                ).value(),
+                scale_type(-pyarrow_object.type.scale),
+                result.c_obj.get().is_valid()
+            )
+        )
+    elif tid == type_id.DECIMAL64:
+        result.c_obj.reset(
+            new fixed_point_scalar[decimal64](
+                (
+                    <fixed_point_scalar[decimal128]*> result.c_obj.get()
+                ).value(),
+                scale_type(-pyarrow_object.type.scale),
+                result.c_obj.get().is_valid()
+            )
+        )
+    elif tid != type_id.DECIMAL128:
+        raise ValueError(
+            "Decimal scalars may only be cast to decimals"
+        )
+
+    return result
+
+
+@from_arrow.register(pa.Array)
+def _from_arrow_column(pyarrow_object, *, DataType data_type=None):
+    if data_type is not None:
+        raise ValueError("data_type may not be passed for arrays")
+    pa_table = pa.table([pyarrow_object], [""])
```

Comment:
Can `pyarrow_unwrap_array` not be used here? https://arrow.apache.org/docs/python/integration/extending.html#pyarrow.pyarrow_unwrap_array

Reply:
We can, but then we don't currently have a libcudf API to process a column. Right now our libcudf API only handles tables (things will be changing once #15047 merges though).

Reply:
Gotcha thanks for the context. A TODO comment here about that might be nice, or I assume you'll probably remember :)

Reply:
I won't forget! We'll have to rewrite this whole module once we finish up that work.